File: android/src/main/java/com/margelo/nitro/nitrovisionkit/utils/VisionKitOptions.kt
package com.margelo.nitro.nitrovisionkit

internal object VisionKitOptions {
  fun trim(options: BackgroundRemovalOptions?) = options?.trim ?: true

  fun retainMask(options: BackgroundRemovalOptions?) = options?.retainMask ?: false

  fun segmentMaxPixels(options: BackgroundRemovalOptions?) =
    options?.maxPixels?.toInt()?.coerceIn(1, VisionKitLimits.MAX_SEGMENT_PIXELS)
      ?: VisionKitLimits.DEFAULT_MAX_PIXELS

  fun maxResults(options: ClassificationOptions?) = options?.maxResults?.toInt() ?: 0

  fun minConfidence(options: ClassificationOptions?) = options?.minConfidence?.toFloat() ?: 0.5f

  fun requireAnalyzeOperations(options: AnalyzeImageOptions) {
    if (options.removeBackground == null && options.classify == null && options.readText == null) {
      throw RuntimeException("analyzeImage requires removeBackground, classify, and/or readText options.")
    }
  }
}

File: ios/Utils/VisionKitOptions.swift
import Foundation
import NitroModules

enum VisionKitOptions {
  static func requireAnalyzeOperations(_ options: AnalyzeImageOptions) throws {
    guard options.removeBackground != nil || options.classify != nil || options.readText != nil else {
      throw RuntimeError("analyzeImage requires removeBackground, classify, and/or readText options.")
    }
  }

  static func trim(_ options: BackgroundRemovalOptions?) -> Bool {
    options?.trim ?? true
  }

  static func retainMask(_ options: BackgroundRemovalOptions?) -> Bool {
    options?.retainMask ?? false
  }

  static func segmentMaxPixels(_ options: BackgroundRemovalOptions?) -> Int {
    let requested = options?.maxPixels ?? Double(VisionKitLimits.defaultMaxPixels)
    guard requested.isFinite else { return VisionKitLimits.defaultMaxPixels }
    let bounded = min(max(requested, 1), Double(VisionKitLimits.maxSegmentPixels))
    return Int(bounded)
  }

  static func maxResults(_ options: ClassificationOptions?) -> Int {
    let requested = options?.maxResults ?? 0
    guard requested.isFinite else { return 0 }
    return Int(min(max(requested, 0), Double(VisionKitLimits.maxClassificationResults)))
  }

  static func minConfidence(_ options: ClassificationOptions?) -> Double {
    guard let c = options?.minConfidence, c.isFinite, c >= 0, c <= 1 else { return 0.5 }
    return c
  }
}
